content/blog/2019-10-23-posh/index.md

```diff
@@ -0,0 +1,315 @@
+it makes sense that *more candidates* for parallelization means *more performance*.
+
+Value prediction makes a big difference in some cases, but less in others;
+but the same theory as before should apply here. Value prediction allows for
+more chances to parallelize, so it should improve performance as long as the
+profiler identifies when it might be a bad idea.
+
+The following graphs shows the importance of using the profiler:
+without it, some programs are _slowed down_ due to the overhead of
+managing tasks.
+The profiler significantly improves performance, and realizes the
+"do no harm" principle of compiler optimization.
+
+<img src="doNoHarm.png" style="width:100%"/>
+
+As we mentioned earlier, most of the speedup comes from executing code in parallel, but
+even when correct parallel execution is not possible and tasks get squashed, memory accesses
+they make have the effect of prefetching data that the re-executions are likely to use.
+To evaluate the impact of prefetching, they modify the simulator so that data brought
+into the processor cache by squashed tasks are marked as invalid.
+Comparing the speedup gained with prefetching and without in the graph before,
+they claim 26% of the speedup is due to prefetching.
+
+<img src="prefetch.png" style="width:100%"/>
+
+
+## Evaluation Takeaways
+
+The paper has a solid evaluation which asks (and answers) all the questions we
+might want asked. Our primary gripe with their evaluation is that it is based on "eyeball statistics".
+No formal null hypothesis testing is done; instead, the authors point at a graph
+and say "the bars are usually higher in this case". Additionally, it's unclear if
+the SPECint benchmarks are really a representative use case for real code. In theory, these
+are meant to test the sequential integer performance of CPUs and may have few opportunities for parallelism.
+On the other hand, the opposite may be true *or* they may represent a good spread of optimizability.
+
+Given the breakdowns that the authors provide, it seems likely that the most significant
+contribution of POSH is its dynamic profiler, which allows their other optimizations
+to be aggressively optimistic. While hardware support does prevent TLS from impacting
+correctness, it doesn't prevent TLS from being a bad idea. The POSH profiler fills this
+gap instead and allows techniques like software value prediction and the structured
+program heuristic to be utilized without hurting performance.
+
+If HTM were actually widely usable by general purpose programs, then POSH would likely
+be an effective optimization for automatically speeding up sequential code!
+
+# Appendix
+
+### Hardware Transactional Memory
+
+Before transactional memory, hardware support for parallel computing
+was limited to synchronization primitives such as [`atomic compare-and-swap`](https://en.wikipedia.org/wiki/Compare-and-swap)
+or [`store-conditional`](https://en.wikipedia.org/wiki/Load-link/store-conditional).
```

Comment:
Not sure you actually want these in Markdown code backticks? They don't really seem like code.